examples/basecoin/app/app.go

```diff
@@ -27,25 +30,60 @@ type BasecoinApp struct {
 	accountMapper sdk.AccountMapper
 }
 
+// NewBasecoinApp - create new BasecoinApp
 // TODO: This should take in more configuration options.
-func NewBasecoinApp() *BasecoinApp {
+// TODO: This should be moved into baseapp to isolate complexity
+func NewBasecoinApp(genesisPath string) *BasecoinApp {
 
 	// Create and configure app.
 	var app = &BasecoinApp{}
-	app.initCapKeys()  // ./init_capkeys.go
-	app.initBaseApp()  // ./init_baseapp.go
-	app.initStores()   // ./init_stores.go
+
+	// TODO open up out of functions, or introduce clarity,
+	// interdependancies are a nightmare to debug
+	app.initCapKeys() // ./init_capkeys.go
+	app.initBaseApp() // ./init_baseapp.go
+	app.initStores()  // ./init_stores.go
+	app.initBaseAppInitStater()
 	app.initHandlers() // ./init_handlers.go
 
-	// TODO: Load genesis
-	// TODO: InitChain with validators
-	// TODO: Set the genesis accounts
+	genesisiDoc, err := bam.GenesisDocFromFile(genesisPath)
+	if err != nil {
+		panic(fmt.Errorf("error loading genesis state: %v", err))
+	}
+
+	// TODO: InitChain with validators from genesis transaction bytes
+
+	// very first begin block used for context when setting genesis accounts
+	header := abci.Header{
+		ChainID:        "",
+		Height:         0,
+		Time:           -1,
+		NumTxs:         -1,
+		LastCommitHash: []byte{0x00},
+		DataHash:       nil,
+		ValidatorsHash: nil,
+		AppHash:        nil,
+	}
+	app.BaseApp.BeginBlock(abci.RequestBeginBlock{
+		Hash:                nil,
+		Header:              header,
```

Comment:
why do we need this ?!

Reply:
to set up the multistore cache wrap for deliver store used in NewContext, I just made this code considerably smaller